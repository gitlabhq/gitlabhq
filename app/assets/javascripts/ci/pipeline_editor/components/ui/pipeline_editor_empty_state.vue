<script>
import { GlButton, GlSprintf, GlEmptyState } from '@gitlab/ui';
import { __ } from '~/locale';
import PipelineEditorFileNav from '~/ci/pipeline_editor/components/file_nav/pipeline_editor_file_nav.vue';
import ExternalConfigEmptyState from '~/ci/common/empty_state/external_config_empty_state.vue';

export default {
  components: {
    GlButton,
    GlSprintf,
    GlEmptyState,
    PipelineEditorFileNav,
    ExternalConfigEmptyState,
  },
  i18n: {
    title: __('Configure a pipeline to automate your builds, tests, and deployments'),
    body: __(
      'Create a %{codeStart}.gitlab-ci.yml%{codeEnd} file in your repository to configure and run your first pipeline.',
    ),
    btnText: __('Configure pipeline'),
  },
  inject: ['emptyStateIllustrationPath', 'usesExternalConfig'],
  methods: {
    createEmptyConfigFile() {
      this.$emit('createEmptyConfigFile');
    },
  },
};
</script>
<template>
  <div>
    <pipeline-editor-file-nav v-on="$listeners" />
    <external-config-empty-state v-if="usesExternalConfig" />

    <gl-empty-state v-else :title="$options.i18n.title" :svg-path="emptyStateIllustrationPath">
      <template #description>
        <gl-sprintf :message="$options.i18n.body">
          <template #code="{ content }">
            <code>{{ content }}</code>
          </template>
        </gl-sprintf>
      </template>
      <template #actions>
        <gl-button
          variant="confirm"
          class="gl-mt-3"
          data-testid="create-new-ci-button"
          @click="createEmptyConfigFile"
        >
          {{ $options.i18n.btnText }}
        </gl-button>
      </template>
    </gl-empty-state>
  </div>
</template>
