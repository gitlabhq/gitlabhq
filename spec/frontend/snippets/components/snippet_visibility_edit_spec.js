import { GlFormRadio, GlIcon, GlFormRadioGroup } from '@gitlab/ui';
import { mount, shallowMount } from '@vue/test-utils';
import SnippetVisibilityEdit from '~/snippets/components/snippet_visibility_edit.vue';
import {
  VISIBILITY_LEVEL_PRIVATE_STRING,
  VISIBILITY_LEVEL_INTERNAL_STRING,
  VISIBILITY_LEVEL_PUBLIC_STRING,
} from '~/visibility_level/constants';
import {
  SNIPPET_VISIBILITY,
  SNIPPET_LEVELS_RESTRICTED,
  SNIPPET_LEVELS_DISABLED,
} from '~/snippets/constants';

describe('Snippet Visibility Edit component', () => {
  let wrapper;
  const defaultVisibilityLevel = 'private';

  function createComponent({
    propsData = {},
    visibilityLevels = [0, 10, 20],
    multipleLevelsRestricted = false,
    deep = false,
  } = {}) {
    const method = deep ? mount : shallowMount;

    wrapper = method.call(this, SnippetVisibilityEdit, {
      propsData: {
        isProjectSnippet: false,
        value: defaultVisibilityLevel,
        ...propsData,
      },
      provide: {
        visibilityLevels,
        multipleLevelsRestricted,
      },
    });
  }

  const findRadios = () => wrapper.findComponent(GlFormRadioGroup).findAllComponents(GlFormRadio);
  const findRadiosData = () =>
    findRadios().wrappers.map((x) => {
      return {
        value: x.find('input').attributes('value'),
        icon: x.findComponent(GlIcon).props('name'),
        description: x.find('.help-text').text(),
        text: x.find('.js-visibility-option').text(),
      };
    });

  describe('rendering', () => {
    it('matches the snapshot', () => {
      createComponent();
      expect(wrapper.element).toMatchSnapshot();
    });

    describe('Visibility options', () => {
      const findRestrictedInfo = () => wrapper.find('[data-testid="restricted-levels-info"]');
      const RESULTING_OPTIONS = {
        0: {
          value: VISIBILITY_LEVEL_PRIVATE_STRING,
          icon: SNIPPET_VISIBILITY.private.icon,
          text: SNIPPET_VISIBILITY.private.label,
          description: SNIPPET_VISIBILITY.private.description,
        },
        10: {
          value: VISIBILITY_LEVEL_INTERNAL_STRING,
          icon: SNIPPET_VISIBILITY.internal.icon,
          text: SNIPPET_VISIBILITY.internal.label,
          description: SNIPPET_VISIBILITY.internal.description,
        },
        20: {
          value: VISIBILITY_LEVEL_PUBLIC_STRING,
          icon: SNIPPET_VISIBILITY.public.icon,
          text: SNIPPET_VISIBILITY.public.label,
          description: SNIPPET_VISIBILITY.public.description,
        },
      };

      it.each`
        levels         | resultOptions
        ${''}          | ${[]}
        ${[]}          | ${[]}
        ${[0]}         | ${[RESULTING_OPTIONS[0]]}
        ${[0, 10]}     | ${[RESULTING_OPTIONS[0], RESULTING_OPTIONS[10]]}
        ${[0, 10, 20]} | ${[RESULTING_OPTIONS[0], RESULTING_OPTIONS[10], RESULTING_OPTIONS[20]]}
        ${[0, 20]}     | ${[RESULTING_OPTIONS[0], RESULTING_OPTIONS[20]]}
        ${[10, 20]}    | ${[RESULTING_OPTIONS[10], RESULTING_OPTIONS[20]]}
      `('renders correct visibility options for $levels', ({ levels, resultOptions }) => {
        createComponent({ visibilityLevels: levels, deep: true });
        expect(findRadiosData()).toEqual(resultOptions);
      });

      it.each`
        levels         | levelsRestricted | resultText
        ${[]}          | ${false}         | ${SNIPPET_LEVELS_DISABLED}
        ${[]}          | ${true}          | ${SNIPPET_LEVELS_DISABLED}
        ${[0]}         | ${true}          | ${SNIPPET_LEVELS_RESTRICTED}
        ${[0]}         | ${false}         | ${''}
        ${[0, 10, 20]} | ${false}         | ${''}
      `(
        'renders correct information about restricted visibility levels for $levels',
        ({ levels, levelsRestricted, resultText }) => {
          createComponent({
            visibilityLevels: levels,
            multipleLevelsRestricted: levelsRestricted,
          });
          expect(findRestrictedInfo().text()).toBe(resultText);
        },
      );

      it('when project snippet, renders special private description', () => {
        createComponent({ propsData: { isProjectSnippet: true }, deep: true });

        expect(findRadiosData()[0]).toEqual({
          value: VISIBILITY_LEVEL_PRIVATE_STRING,
          icon: SNIPPET_VISIBILITY.private.icon,
          text: SNIPPET_VISIBILITY.private.label,
          description: SNIPPET_VISIBILITY.private.description_project,
        });
      });

      it('when project snippet, renders special public description', () => {
        createComponent({ propsData: { isProjectSnippet: true }, deep: true });

        expect(findRadiosData()[2]).toEqual({
          value: VISIBILITY_LEVEL_PUBLIC_STRING,
          icon: SNIPPET_VISIBILITY.public.icon,
          text: SNIPPET_VISIBILITY.public.label,
          description: SNIPPET_VISIBILITY.public.description_project,
        });
      });
    });
  });

  describe('functionality', () => {
    it('pre-selects correct option in the list', () => {
      const value = VISIBILITY_LEVEL_INTERNAL_STRING;

      createComponent({ propsData: { value } });

      expect(wrapper.findComponent(GlFormRadioGroup).attributes('checked')).toBe(value);
    });
  });
});
