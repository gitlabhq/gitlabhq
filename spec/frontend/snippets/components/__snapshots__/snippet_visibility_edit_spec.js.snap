// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snippet Visibility Edit component rendering matches the snapshot 1`] = `
<div
  class="form-group"
>
  <label>
    Visibility level
  </label>
  <gl-form-group-stub
    class="gl-mb-0"
    id="reference-0"
    labeldescription=""
    optionaltext="(optional)"
  >
    <gl-form-radio-group-stub
      checked="private"
      disabledfield="disabled"
      htmlfield="html"
      options=""
      stacked=""
      textfield="text"
      valuefield="value"
    >
      <gl-form-radio-stub
        class="gl-mb-3"
        value="private"
      >
        <div
          class="gl-flex gl-items-center"
        >
          <gl-icon-stub
            name="lock"
            size="16"
            variant="current"
          />
          <span
            class="gl-font-semibold gl-ml-2 js-visibility-option"
            data-qa-visibility="Private"
            data-testid="visibility-content"
          >
            Private
          </span>
        </div>
      </gl-form-radio-stub>
      <gl-form-radio-stub
        class="gl-mb-3"
        value="internal"
      >
        <div
          class="gl-flex gl-items-center"
        >
          <gl-icon-stub
            name="shield"
            size="16"
            variant="current"
          />
          <span
            class="gl-font-semibold gl-ml-2 js-visibility-option"
            data-qa-visibility="Internal"
            data-testid="visibility-content"
          >
            Internal
          </span>
        </div>
      </gl-form-radio-stub>
      <gl-form-radio-stub
        class="gl-mb-3"
        value="public"
      >
        <div
          class="gl-flex gl-items-center"
        >
          <gl-icon-stub
            name="earth"
            size="16"
            variant="current"
          />
          <span
            class="gl-font-semibold gl-ml-2 js-visibility-option"
            data-qa-visibility="Public"
            data-testid="visibility-content"
          >
            Public
          </span>
        </div>
      </gl-form-radio-stub>
    </gl-form-radio-group-stub>
  </gl-form-group-stub>
  <div
    class="gl-text-subtle"
    data-testid="restricted-levels-info"
  />
</div>
`;
